Filled Menu State

clayjohn opened the pull request:
I have added most of what I wanted in the menu. Let me know if there is other stuff we should put in there. I tried to make it as easy as possible to add new functionality.

clayjohn committed:
Merge remote-tracking branch 'refs/remotes/CalebJohn/master'

clayjohn committed:
Merge remote-tracking branch 'refs/remotes/CalebJohn/master'

clayjohn committed:
Added content to Menu State

Content includes: Background shader, title, buttons, and options menu

CalebJohn commented:
So a couple of things here, first take a look at [this](http://phaser.io/examples/v2/filters/multiple-shaders) for a way to define shaders in a new file as opposed to having them in the javascript source files (btw shaders should go in a new directory, 'assets/shaders' maybe?).

I'm also not a fan of what you've done in the preload function, I think var bindings are local to functions in javascript so the lines you have there aren't doing anything. The variables being defined in `create`, etc are actually being bound as global variables, which is bad. Take a look at what @enochtsang has done in  #12 for state variables (They should be accessed by `this`  and declared in that top function).

And as a final note we should have some utility scripts for stuff like buttons so that we don't need helper functions like the one at line 68. You don't need to worry about that now, but we'll need to make sure that this code gets rewritten at some point to include that.

And to end on a light note, this looks fantastic! I'm really happen with it.

clayjohn commented:
Okay I actually spent a long time trying to figure out the variable definitions, so thanks for that I'll fix it up. Is there a way to make my changes and then update this pull request or do I make a new one?

CalebJohn commented:
Make the changes on your branch like normal, commit like normal and push to your origin branch, the pull will be automatically updated.

clayjohn committed:
Various fixes after the rebase

Cleaned up all kinds of things

Caleb's suggested changes

Initialized variables to null. Added semicolon to line 62. removed comma
in line 123.

Added "use strict" for js hint and fixed typo

Changed TOD in shader code

Changed to dTime as in daytime

Updated references to 'game'

I changed all references to game to window.game in order to facilitate
jsHint. Im not sure this is the best solution but it works for now.

added globals window

fixed whitespace issues

clayjohn committed:
Cleaned up variables and Shader definition

Changed all variables to be declared locally. And moved the shader
creation off to a new script under assests/shader

clayjohn committed:
Merge remote-tracking branch 'refs/remotes/origin/MainMenu'

enochtsang commented:
I am getting the error: 
```
127.0.0.1:57695 [404]: /assets/phaser.map - No such file or directory
```
Can you include that file in this pull request?

clayjohn commented:
Just looked into that, it seems to be an internal Phaser error. http://www.html5gamedevs.com/topic/9151-ask-error-get-phasermap/
I'm not getting the same error when I run it.

CalebJohn commented:
I fixed that error in my upcoming pull request by adding phaser.js instead of phaser.min.js. It's nothing to worry about here.

CalebJohn commented:
@clayjohn Theres a couple more small issues to fix before we merge this,
line 62 is missing a semicolon,
you don't need the comma at the end of line 123,
and all the definitions at the beginning should probably be set to null, or whatever is appropriate.

Once those are fixed we can merge this in!

clayjohn committed:
Caleb's suggested changes

Initialized variables to null. Added semicolon to line 62. removed comma
in line 123.

clayjohn commented:
Made the changes.

enochtsang commented:
Your travis build is still failing due to JsHint issues. There are some tricks to getting some of JsHints  errors, let me know if you need help with that. You can also see `WorldGenerator.js` and `Cutscene.js` to  for how to define global variables for JsHint.

ExtraGravity commented on assets/shaders/MenuShader.js:
```diff
@@ -0,0 +1,98 @@
+    		    "uv.y/=resolution.y;",
+    		    "uv.x*=(resolution.x/resolution.y);",
+    		    
+    		    "//TOD calculation",
```
Typo here, probably worth to fix so we can CTRL +F TODO later

clayjohn replied:
TOD stands for Time of Day, not sure what TODO would be.

CalebJohn replied:
TODO as in, to do list.

ExtraGravity replied:
Oh haha, yea disregard my comment then haha.

clayjohn replied:
I changed it anyways. Just so it doesn't clash when we are looking for our to do lists.

clayjohn committed:
Added "use strict" for js hint and fixed typo

clayjohn committed:
Changed TOD in shader code

Changed to dTime as in daytime

clayjohn committed:
Updated references to 'game'

I changed all references to game to window.game in order to facilitate
jsHint. Im not sure this is the best solution but it works for now.

clayjohn commented:
I updated all references to game be to window.game because that is what was causing the conflicts in jshint. but apparently it doesn't like references to window either when it runs in travis. I don't get any errors when I check in sublime. (except for not recognizing Phaser of course)

CalebJohn commented:
To get jshint to work, you're going to need remove the `"use strict"` and add `/* globals window */` at the top.

clayjohn commented:
I just tried that and now Jshint throws an error on every single line and says i'm missing the use strict statement. Should I maybe have both?

clayjohn committed:
added globals window

clayjohn committed:
fixed whitespace issues

clayjohn committed:
Merge branch 'MainMenu' of https://github.com/clayjohn/potential-fortnight into MainMenu

ExtraGravity commented on src/states/Menu.js:
```diff
@@ -1,20 +1,126 @@
+    window.game.stage.backgroundColor = '#FFFFFF';
+    window.game.scale.fullScreenScaleMode = Phaser.ScaleManager.NO_SCALE;
+    //load Filter
+    //this.initMenuBackground();
```
These two comments should be removed to reduce clutter.

ExtraGravity commented on src/states/Menu.js:
```diff
@@ -1,20 +1,126 @@
+    window.game.scale.fullScreenScaleMode = Phaser.ScaleManager.NO_SCALE;
+    //load Filter
+    //this.initMenuBackground();
+    //Apply filter to image in background
```
Personally I think using 
`/* Apply filter to image in background */`
looks much nicer for "section" comments, I would use `//` for writing comments beside a specific line. That's my personal preference cause I think it looks alot nicer, up to you if you want to change that though.

ExtraGravity commented on src/states/Menu.js:
```diff
@@ -1,20 +1,126 @@
+        window.game.scale.startFullScreen();
+        
+        
+    }
```
Please fix the whitespacing in this function.

ExtraGravity commented on src/states/Menu.js:
```diff
@@ -1,20 +1,126 @@
+    this.muteButton.visible = this.muteButton.visible === false;
+    this.seedButton.visible = this.seedButton.visible === false;
+    this.returnButton.visible = this.returnButton.visible === false;
+  },
```
Clever! It's pretty, but I think it has clarity and scaling issues. 

This kind of only accounts for two states for the main menu, I think using a state machine where a state is kept track of like `OPTIONS` and `MAIN` then using an if statement to show certain buttons would be clearer and scale much better.

enochtsang commented:
That's it for my comments, I'll be fine with merging it once the issues I commented on are fixed.

clayjohn committed:
Updated options menu plus minor tweaks to comments and whitespace

Options menu now is slightly more robust as it uses states instead of a
simple toggle.

clayjohn commented:
Okay. I made your changes. I set up a fairly simple state system to handle the states for the menu, it should be easy to have a few different states if we decide we want that. Let me know if you want me to add something even more robust.

enochtsang commented:
For the state system, `switch case` statements would be more suitable than `if else` statements, switch statements are also a common feature in other languages. 

More info here: http://www.w3schools.com/js/js_switch.asp

But it doesn't matter for the prototype, merging.
